Sources/BSON/Integers+BSONValue.swift

```diff
@@ -1,6 +1,34 @@
 import NIO
 
 extension Int32: BSONValue {
+    /// Initializes an `Int32` given well-formatted canonical or relaxed extended JSON representing an `Int32`
+    /// Returns `nil` if the provided value is not an Int32.
+    /// Throws if the JSON is a partial match or is malformed.
+    internal init?(fromExtJSON json: JSON) throws {
+        switch json {
+        case let .number(n):
+            // relaxed extended JSON
+            guard let int = Int32(exactly: n) else {
+                return nil
+            }
+            self = int
+        case let .object(obj):
+            // canonical extended JSON
+            guard obj.count == 1 else {
+                throw BSONError.InternalError(message: "Not a valid Int32")
+            }
+            guard let str = obj["$numberInt"]?.stringValue else {
+                throw BSONError.InternalError(message: "Not a valid Int32")
+            }
+            guard let int = Int32(str) else {
+                throw BSONError.InternalError(message: "Not a valid Int32")
```

Comment:
These errors will be thrown by `ExtendedJSONDecoder`, so they need to be `DecodingError`. `BSONError.InternalError` is reserved basically for invariant failures or other unexplainable errors, usually indicating bugs in our code.

Reply:
Hmm, good point. I think `[]` as the `codingPath` makes sense, since this error is occurring during preprocessing. We probably will want to include the key we're at in the error message though, so I think this initializer should have its own `keyPath: [String]` that gets built up as you recursively descend into the BSON structure (more recent keys being appended to the path as we go down).

Then the error messages could include something like: "doc.subdoc.array.1: could not parse Int32 from "bladfljk"".

Reply:
keyPath would only get set through the extended json initializers on object and array right? 

Would we want it to be set for the cases like `["$numberInt":"bla"]` and have the error there be `"$numberInt: could not parse Int32 from "bla""`?

Reply:
>keyPath would only get set through the extended json initializers on object and array right?

Yep, exactly.

>Would we want it to be set for the cases like ["$numberInt":"bla"] and have the error there be "$numberInt: could not parse Int32 from "bla""?

Nope, those keys are really part of the "value" of the integer.

Reply:
okay I added this and added a test that passes one in. should I try to test that the debug message is as expected? or is that unnecessary

Reply:
I think that's unnecessary for now, but we can do that for the `Document` / `Array` cases.